Add the encrypted storage APIs

mosic opened the pull request:
This adds four new calls to the SDK:
* `sdk.storage.set()`
* `sdk.storage.get()`
* `sdk.storage.list()`
* `sdk.storage.promptUpload()`

All files/string uploaded this way are streams pushed to Swarm, encrypted with libsodium.

mosic commented on packages/sdk/src/apis/Storage.js:
```diff
@@ -0,0 +1,21 @@
+import ClientAPIs from '../ClientAPIs'
+
+export default class StorageAPIs extends ClientAPIs {
+  promptUpload(params: { name: string }): Promise<?string> {
```
Yeah, that makes sense to me. I changed the `name` argument to `key`.

mosic commented:
@PaulLeCam Thanks for the detailed review! I agree with the proposed changes, I was aware that the PR was a bit rushed.

As for the additional APIs to remove an entry and prompting to download/save a file, could that be in a separate PR?

Either way, I'll work more on this tomorrow, for now I pushed the changes that are tested and work.

PaulLeCam commented:
Thanks for the changes!
Yes no problem having these 2 APIs as a separate PR, I don't think we need to provide these in the next release anyways so no rush adding them.

mosic commented:
@PaulLeCam I addressed all your comments, let me know if I can do something better instead of raising errors when I'm validating the storage requests.

PaulLeCam commented on packages/launcher/src/main/rpc/app.js:
```diff
@@ -31,9 +31,7 @@ class CommsSubscription extends ContextSubscription<RxSubscription> {
-const getStorageManifestHash = async (
-  ctx: AppContext,
-): Promise<Array<string>> => {
+const getStorageManifestHash = async (ctx: AppContext): Promise<Object> => {
```
Returned type could be explicitly set to `Promise<{ feedHash: string, manifestHash: string }>` as the shape is known, it's more specific than `Object`.

mosic replied:
Nice, I didn't know that you could specify keys in the object for the type :+1:

PaulLeCam commented on packages/launcher/src/main/rpc/app.js:
```diff
@@ -53,6 +51,15 @@ const getStorageManifestHash = async (
 }
 
+const raiseErrorIfKeyInvalid = (key: string): void | Error => {
+  const validFilename = /^([A-Za-z0-9_. =+-]+?)$/
```
Rather than redefining this regex in every function call, you could set it at the module level so it's only defined once.

mosic replied:
Good point :+1:

PaulLeCam commented on packages/launcher/src/main/rpc/app.js:
```diff
@@ -53,6 +51,15 @@ const getStorageManifestHash = async (
   }
 }
 
+const raiseErrorIfKeyInvalid = (key: string): void | Error => {
```
If the error is thrown, the returned type here should only be `void`, as throwing an error is an exception.
This return type would work if `return new Error()` was used rather than `throw new Error()`.

PaulLeCam commented on packages/launcher/src/main/rpc/app.js:
```diff
@@ -270,9 +280,9 @@ export const sandboxed = {
-          if (list.entries) {
-            entries = list.entries.map(meta => {
+          const response = await ctx.bzz.list(contentHash)
+          if (response.entries instanceof Array) {
```
Does `instanceof Array` work as expected here? AFAIK `Array.isArray()` should be used to type-check arrays.

mosic replied:
It does, but that's even better, thanks :+1:

mosic replied:
Hm, flow complains if I use that check instead:

```
$ flow check --merge-timeout 0
Error ┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈ packages/launcher/src/main/rpc/app.js:286:34

Cannot call await with ctx.bzz.list(...) bound to p because property entries of unknown type [1] is incompatible with
ListEntry [2].

     packages/launcher/src/main/rpc/app.js
     283│             { mode: 'content-hash' },
     284│           )
     285│           ctx.storage.contentHash = contentHash
     286│           const response = await ctx.bzz.list(contentHash)
 [1] 287│           if (Array.isArray(response.entries)) {
     288│             entries = response.entries.map(meta => {
     289│               return { contentType: meta['contentType'], key: meta['path'] }

     node_modules/@erebos/api-bzz-base/cjs/types.js.flow
 [2]  26│   entries?: Array<ListEntry>,
```
